### PR TITLE
Propagator batching for the file scanner

### PR DESCRIPTION
**Pull request by icewind1991**
Batch propagator updates in the file scanner for a [large performance improvements](https://blackfire.io/profiles/compare/2d8e4020-3968-4d14-b182-e0f8c5d4b476/graph) when scanning a folder with a lot of changes

cc @PVince81

### CODE REVIEW

File: lib/public/Files/Cache/IPropagator.php [hunk 1]
```diff
@@ -28,6 +28,21 @@
...
 interface IPropagator {
 	/**
+	 * Mark the beginning of a propagation batch
+	 *
+	 * Note that not all cache setups support propagation in which case this will be a noop
+	 *
+	 * Batching for cache setups that do support it has to be explicit since the cache state is not fully consistent
+	 * before the batch is committed.
+	 */
+	public function beginBatch();
```
Comment (PVince81) [inline lib/public/Files/Cache/IPropagator.php]: add `@since` attribute.

bad thing that this is changing an interface post feature freeze

**Comment by PVince81** [comment]
Great addition.

Are there any use cases where batching does not make sense ?

**Comment by DeepDiver1975** [comment]
:boom: 

````
06:14:33 Using PHP executable /opt/phpenv/shims/php
06:14:34 Parsing all files in lib/public for the presence of @since or @deprecated on each method...
06:14:34 
06:14:34 @since or @deprecated tag is needed in PHPDoc for OCP\Files\Cache\IPropagator::beginBatch
06:14:34 @since or @deprecated tag is needed in PHPDoc for OCP\Files\Cache\IPropagator::commitBatch

**Comment by PVince81** [comment]
Please add unit tests.
Such complex logic and no tests make me shiver...

**Comment by PVince81** [comment]
Moving back to "Developing" as this PR isn't complete and tests fail

File: lib/private/Files/Cache/Propagator.php [hunk 2]
```diff
@@ -98,4 +109,79 @@ protected function getParents($path) {
...
+			}
+		}
+	}
+
+	/**
+	 * Commit the active propagation batch
+	 */
+	public function commitBatch() {
+		if (!$this->inBatch) {
+			return;
```
Comment (nickvergessen) [inline lib/private/Files/Cache/Propagator.php]: throw an exception to make the coding error visible?

File: tests/lib/Files/Cache/PropagatorTest.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+use Test\TestCase;
+
+/**
+ * @group DB
+ */
+class PropagatorTest extends TestCase {
+	/** @var  IStorage */
+	private $storage;
+
+	public function setUp() {
```
Comment (nickvergessen) [inline tests/lib/Files/Cache/PropagatorTest.php]: missing `parent::setUp();`

**Comment by nickvergessen** [comment]
> Test\Files\Cache\PropagatorTest::testTimePropagation
> Failed asserting that 1464871257 matches expected 1464871357.
>
> /ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/tests/lib/Files/Cache/PropagatorTest.php:68

File: tests/lib/Files/Cache/PropagatorTest.php (hunk 3 above)
Comment (icewind1991) [inline tests/lib/Files/Cache/PropagatorTest.php]: fixed

**Comment by icewind1991** [comment]
Fixed tests

**Comment by icewind1991** [comment]
@owncloud/filesystem this is ready to review

**Comment by PVince81** [comment]
@icewind1991 does this also affect the regular scanner / update detection for ext storage ?
Or only background scan ? (just to find out what needs to be retested)

**Comment by icewind1991** [comment]
Only scans run from `occ` atm

**Comment by PVince81** [comment]
The code looks fine, @SergioBertolinSG can you help testing ?

**Comment by SergioBertolinSG** [comment]
When listening to changes on windows network drive I get an exception. The files are correctly synced though.
 
[Exception]            
  Invalid configuration 

```
{"reqId":"2HiN0\/mlcMsucrWF4cMg","remoteAddr":"","app":"PHP","message":"Undefined index: user at \/opt\/owncloud\/apps\/windows_network_drive\/lib\/wnd.php#42","level":3,"time":"2016-06-06T13:06:01+00:00","method":"--","url":"--","user":"--"}
{"reqId":"fa661eZcxn2a0WLm5GCh","remoteAddr":"","app":"PHP","message":"Undefined index: user at \/opt\/owncloud\/apps\/windows_network_drive\/lib\/wnd.php#42","level":3,"time":"2016-06-06T13:40:33+00:00","method":"--","url":"--","user":"--"}
{"reqId":"6HD\/t2tc7n6oqN3j9JSO","remoteAddr":"","app":"PHP","message":"Undefined index: user at \/opt\/owncloud\/apps\/windows_network_drive\/lib\/wnd.php#42","level":3,"time":"2016-06-06T13:54:39+00:00","method":"--","url":"--","user":"--"}
{"reqId":"3P3TdxKDrnPF2Df+E3tR","remoteAddr":"","app":"PHP","message":"Undefined index: user at \/opt\/owncloud\/apps\/windows_network_drive\/lib\/wnd.php#42","level":3,"time":"2016-06-06T13:55:06+00:00","method":"--","url":"--","user":"--"}
```

**Comment by PVince81** [comment]
@icewind1991 any update ?

Also the "Invalid configuration" somewhat looks like what @felixboehm posted.

@SergioBertolinSG any details about the SMB setup ?

**Comment by SergioBertolinSG** [comment]
> any details about the SMB setup ?

It is a windows server 2008.

**Comment by PVince81** [comment]
@SergioBertolinSG details about the SMB configuration. Which authentication mode ?

**Comment by SergioBertolinSG** [comment]
Ah sorry, that was using global credentials.

**Comment by icewind1991** [comment]
The wnd errors are unrelated to this PR

**Comment by PVince81** [comment]
@SergioBertolinSG can you retest with non-global credentials ?

**Comment by SergioBertolinSG** [comment]
> @SergioBertolinSG can you retest with non-global credentials ?

Yes, I've tested this using login credentials, save in database. When renaming a file I get the 

```
[Exception]            
  Invalid configuration  
```

**Comment by PVince81** [comment]
Ok, so there's a problem with renames

**Comment by icewind1991** [comment]
Try using normal credentials (where you specify the password directly in the external storage config) all other credential mechanisms suffer from the same problem

**Comment by SergioBertolinSG** [comment]
Using username and password credentials it works fine without exceptions.

**Comment by PVince81** [comment]
The sad part: when running these two failing tests alone they pass... there's some side effect probably.

Anyway, this PR already fixes one test, please review @owncloud/filesystem

**Comment by PVince81** [comment]
Propagation still works :+1: 

Tested with SFTP and manually setting size=-1 in DB folders and running `occ files:scan --all --unscanned`

**Comment by PVince81** [comment]
Counting https://github.com/owncloud/core/pull/24863#issuecomment-225177985 as :+1:

**Pull request merged**